blocklylib-core/src/main/java/com/google/blockly/android/codegen/CodeGeneratorManager.java

```diff
@@ -43,54 +43,41 @@ public void onServiceDisconnected(ComponentName arg0) {
         };
     }
 
-    public void onResume() {
-        Intent intent = new Intent(mContext, CodeGeneratorService.class);
-        mContext.bindService(intent, mCodeGenerationConnection, Context.BIND_AUTO_CREATE);
-    }
-
+    /**
+     * Unbind the underlying service (if it is bound).
+     */
     public void onPause() {
-        mContext.unbindService(mCodeGenerationConnection);
+        if (isBound()) {
+            mContext.unbindService(mCodeGenerationConnection);
+        }
     }
 
     /**
      * Calls the Service to request code generation for the workspace passed in.
      *
-     * @param workspace the workspace to request code generation for.
-     * @param codeGenerationCallback the callback to call with the generated code.
+     * @param codeGenerationRequest the request to generate code.
      */
-    public void requestCodeGeneration(List<String> blockDefinitionsJsonPaths,
-                                      List<String> generatorsJsPaths,
-                                      Workspace workspace,
-                                      CodeGeneratorCallback codeGenerationCallback) {
-
-        if (workspace.hasBlocks()) {
-            if (isBound()) {
-                try {
-                    final StringOutputStream serialized = new StringOutputStream();
-                    workspace.serializeToXml(serialized);
-
-                    mGeneratorService.requestCodeGeneration(
-                        new CodeGenerationRequest(
-                            serialized.toString(),
-                            codeGenerationCallback,
-                            blockDefinitionsJsonPaths,
-                            generatorsJsPaths)
-                    );
-                } catch (BlocklySerializerException e) {
-                    Log.wtf(TAG, e);
-                    Toast.makeText(mContext, "Code generation failed.",
-                        Toast.LENGTH_LONG).show();
-
-                }
-            } else {
-                Log.i(TAG, "Generator not bound to service. Skipping run request.");
-            }
+    public void requestCodeGeneration(CodeGenerationRequest codeGenerationRequest) {
+        if (isBound()) {
+            executeCodeGenerationRequest(codeGenerationRequest);
         } else {
-            Log.i(TAG, "No blocks in workspace. Skipping run request.");
+            mStoredRequests.add(codeGenerationRequest);
+            connectToService();
         }
     }
 
     private boolean isBound() {
         return mGeneratorService != null;
     }
+
+    private void connectToService() {
+        Intent intent = new Intent(mContext, CodeGeneratorService.class);
```

Comment:
Add a mIsConnecting flag. Set it to true after bindService is called and false when the service connects. In this method check if (!isBound() && !mIsConnecting) before calling bindService.